### PR TITLE
fix #282: datetime column in associate table based on connection dialect

### LINKED ISSUE #282: Datetime column in assoc table with sqlite DB does not work
When I use a datetime column with `default` in my assoc table in sqlite - sqlalchemy_continuum is throwing an error in `append_association_operation`

Reproducible example:
```python
import datetime

from sqlalchemy import Column, ForeignKey, Integer, DateTime, Table, create_engine, func, String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker, relationship, configure_mappers
from sqlalchemy_continuum import make_versioned


make_versioned(user_cls=None)

Base = declarative_base()


book_author_table = Table(
    'book_author',
    Base.metadata,
    Column('book_id', Integer, ForeignKey('book.id'), primary_key=True, nullable=False),
    Column('author_id', Integer, ForeignKey('author.id'), primary_key=True, nullable=False),
    Column('created_date', DateTime, nullable=False, server_default=func.current_timestamp(), default=datetime.datetime.utcnow),
)


class Book(Base):
    __tablename__ = 'book'
    __versioned__ = {}

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String, nullable=False)
    authors = relationship('Author', secondary=book_author_table, back_populates='books')


class Author(Base):
    __tablename__ = 'author'
    __versioned__ = {}

    id = Column(Integer, primary_key=True, autoincrement=True)
    name = Column(String, nullable=False)
    books = relationship('Book', secondary=book_author_table, back_populates='authors')

configure_mappers()
engine = create_engine('sqlite:///db.sqlite', echo=True)

# Create all tables
Base.metadata.create_all(bind=engine)

# Create a session
session = sessionmaker()
session.configure(bind=engine)
db = session()


lotr = Book(name='Lord of the rings')
tolkien = Author(name='JRR Tolkien', books=[lotr])
db.add(lotr)
db.add(tolkien)
db.commit()
```

To run this, you can do:
```
$ python -m venv venv
$ venv/bin/pip install sqlalchemy==1.4.39 sqlalchemy-continuum==1.3.12
$ venv/bin/python app.py
BEGIN (implicit)
PRAGMA main.table_info("book_author")
[raw sql] ()
PRAGMA temp.table_info("book_author")
[raw sql] ()
PRAGMA main.table_info("book")
[raw sql] ()
PRAGMA temp.table_info("book")
[raw sql] ()
PRAGMA main.table_info("author")
[raw sql] ()
PRAGMA temp.table_info("author")
[raw sql] ()
PRAGMA main.table_info("book_version")
[raw sql] ()
PRAGMA temp.table_info("book_version")
[raw sql] ()
PRAGMA main.table_info("author_version")
[raw sql] ()
PRAGMA temp.table_info("author_version")
[raw sql] ()
PRAGMA main.table_info("transaction")
[raw sql] ()
PRAGMA temp.table_info("transaction")
[raw sql] ()
PRAGMA main.table_info("book_author_version")
[raw sql] ()
PRAGMA temp.table_info("book_author_version")
[raw sql] ()

CREATE TABLE book (
        id INTEGER NOT NULL,
        name VARCHAR NOT NULL,
        PRIMARY KEY (id)
)


[no key 0.00008s] ()

CREATE TABLE author (
        id INTEGER NOT NULL,
        name VARCHAR NOT NULL,
        PRIMARY KEY (id)
)


[no key 0.00015s] ()

CREATE TABLE book_version (
        id INTEGER NOT NULL,
        name VARCHAR,
        transaction_id INTEGER NOT NULL,
        end_transaction_id INTEGER,
        operation_type SMALLINT NOT NULL,
        PRIMARY KEY (id, transaction_id)
)


[no key 0.00019s] ()
CREATE INDEX ix_book_version_transaction_id ON book_version (transaction_id)
[no key 0.00017s] ()
CREATE INDEX ix_book_version_operation_type ON book_version (operation_type)
[no key 0.00021s] ()
CREATE INDEX ix_book_version_end_transaction_id ON book_version (end_transaction_id)
[no key 0.00017s] ()

CREATE TABLE author_version (
        id INTEGER NOT NULL,
        name VARCHAR,
        transaction_id INTEGER NOT NULL,
        end_transaction_id INTEGER,
        operation_type SMALLINT NOT NULL,
        PRIMARY KEY (id, transaction_id)
)


[no key 0.00022s] ()
CREATE INDEX ix_author_version_transaction_id ON author_version (transaction_id)
[no key 0.00018s] ()
CREATE INDEX ix_author_version_end_transaction_id ON author_version (end_transaction_id)
[no key 0.00018s] ()
CREATE INDEX ix_author_version_operation_type ON author_version (operation_type)
[no key 0.00017s] ()

CREATE TABLE "transaction" (
        issued_at DATETIME,
        id INTEGER NOT NULL,
        remote_addr VARCHAR(50),
        PRIMARY KEY (id)
)


[no key 0.00015s] ()

CREATE TABLE book_author_version (
        book_id INTEGER NOT NULL,
        author_id INTEGER NOT NULL,
        created_date DATETIME DEFAULT (CURRENT_TIMESTAMP),
        transaction_id INTEGER NOT NULL,
        end_transaction_id INTEGER,
        operation_type SMALLINT NOT NULL,
        PRIMARY KEY (book_id, author_id, transaction_id)
)


[no key 0.00017s] ()
CREATE INDEX ix_book_author_version_transaction_id ON book_author_version (transaction_id)
[no key 0.00014s] ()
CREATE INDEX ix_book_author_version_operation_type ON book_author_version (operation_type)
[no key 0.00014s] ()
CREATE INDEX ix_book_author_version_end_transaction_id ON book_author_version (end_transaction_id)
[no key 0.00014s] ()

CREATE TABLE book_author (
        book_id INTEGER NOT NULL,
        author_id INTEGER NOT NULL,
        created_date DATETIME DEFAULT (CURRENT_TIMESTAMP) NOT NULL,
        PRIMARY KEY (book_id, author_id),
        FOREIGN KEY(book_id) REFERENCES book (id),
        FOREIGN KEY(author_id) REFERENCES author (id)
)


[no key 0.00020s] ()
COMMIT
BEGIN (implicit)
INSERT INTO "transaction" (issued_at, remote_addr) VALUES (?, ?)
[generated in 0.00036s] ('2022-08-17 05:13:58.336324', None)
INSERT INTO author (name) VALUES (?)
[generated in 0.00021s] ('JRR Tolkien',)
INSERT INTO book (name) VALUES (?)
[generated in 0.00020s] ('Lord of the rings',)
INSERT INTO book_author (book_id, author_id, created_date) VALUES (?, ?, ?)
[generated in 0.00053s] (1, 1, '2022-08-17 05:13:58.342296')
ROLLBACK
Traceback (most recent call last):
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1720, in _execute_context
    dialect, self, conn, execution_options, *args, **kw
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 1076, in _init_compiled
    for key in positiontup
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 1076, in <listcomp>
    for key in positiontup
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/dialects/sqlite/base.py", line 1004, in process
    "SQLite DateTime type only accepts Python "
TypeError: SQLite DateTime type only accepts Python datetime and date objects as input.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "app.py", line 61, in <module>
    db.commit()
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 1451, in commit
    self._transaction.commit(_to_root=self.future)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 829, in commit
    self._prepare_impl()
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 808, in _prepare_impl
    self.session.flush()
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3383, in flush
    self._flush(objects)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3523, in _flush
    transaction.rollback(_capture_exception=True)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 72, in __exit__
    with_traceback=exc_tb,
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3487, in _flush
    self.dispatch.after_flush(self, flush_context)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/event/attr.py", line 256, in __call__
    fn(*args, **kw)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy_continuum/manager.py", line 354, in after_flush
    uow.process_after_flush(session)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy_continuum/unit_of_work.py", line 95, in process_after_flush
    self.make_versions(session)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy_continuum/unit_of_work.py", line 308, in make_versions
    self.create_association_versions(session)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy_continuum/unit_of_work.py", line 295, in create_association_versions
    session.execute(stmt)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 1712, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1631, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 333, in _execute_on_connection
    self, multiparams, params, execution_options
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1508, in _execute_clauseelement
    cache_hit=cache_hit,
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1726, in _execute_context
    e, util.text_type(statement), parameters, None, None
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2044, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 208, in raise_
    raise exception
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1720, in _execute_context
    dialect, self, conn, execution_options, *args, **kw
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 1076, in _init_compiled
    for key in positiontup
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 1076, in <listcomp>
    for key in positiontup
  File "/home/abdealijk/venv/lib/python3.7/site-packages/sqlalchemy/dialects/sqlite/base.py", line 1004, in process
    "SQLite DateTime type only accepts Python "
sqlalchemy.exc.StatementError: (builtins.TypeError) SQLite DateTime type only accepts Python datetime and date objects as input.
[SQL: INSERT INTO book_author_version (book_id, author_id, created_date, transaction_id, operation_type) VALUES (?, ?, ?, ?, ?)]
```

### PR DESCRIPTION
**Pull request by indiVar0508**
attempt to address #282, to process the input parameters received in @track_association_operations method

before changes Pytest results
346 failed, 308 passed, 14 skipped, 5398 warnings, 90 errors
After Changes Pytest results
346 failed, 308 passed, 14 skipped, 5398 warnings, 90 errors

### CODE REVIEW

**Comment by indiVar0508** [comment]
Hi @marksteward ,

Did you get a chance to review this PR ?

**Comment by marksteward** [comment]
Not yet.

File: sqlalchemy_continuum/manager.py [hunk 1]
```diff
@@ -431,6 +431,21 @@ def track_cloned_connections(self, c, opt):
...
 
+    def get_python_processed_parameter(self, parameter, column, dialect):
+        param_transformer = column.type.dialect_impl(dialect)
+        process = param_transformer.result_processor(dialect, column.type)
+        if process is not None:
+            # https://github.com/sqlalchemy/sqlalchemy/blob/231262d4782fc4d1a5d5fdbf985a06e04270e878/lib/sqlalchemy/sql/type_api.py#L452
+            return process(parameter)
+        return parameter
+
+    def process_parameters(self, table_name, parameters, dialect):
```
Comment (AbdealiJK) [inline sqlalchemy_continuum/manager.py:442]: (Opinionated) We don't really need 2 functions here - we can do it in a single function with a for loop ?

File: sqlalchemy_continuum/manager.py [hunk 2]
```diff
@@ -431,6 +431,21 @@ def track_cloned_connections(self, c, opt):
...
+        table = self.metadata.tables[table_name]
+        return tuple(
+            self.get_python_processed_parameter(parameter=parameters[idx], column=column, dialect=dialect) 
+            for idx, column in enumerate(table.columns)
```
Comment (AbdealiJK) [inline sqlalchemy_continuum/manager.py:446]: I am not sure I get this.
You are doing a `enumerate(table.columns)` - so lets say I have 10 columns in my table.
Not you do `parameters[idx]` - You are assuming every parameter corresponds to each column ?

There are some asusmptions you seem to be making:
1. This is only valid for INSERT statements
2. Are we sure that ALL columns will be present ? What if there is a `server_default` or nullabe=True which makes the column null by default ?

File: sqlalchemy_continuum/manager.py [hunk 3]
```diff
@@ -431,6 +431,21 @@ def track_cloned_connections(self, c, opt):
                 if not connection.closed and connection.connection is c.connection:  # ConnectionFairy is the same - this is a clone
                     self.units_of_work[c] = uow
 
+    def get_python_processed_parameter(self, parameter, column, dialect):
+        param_transformer = column.type.dialect_impl(dialect)
+        process = param_transformer.result_processor(dialect, column.type)
+        if process is not None:
+            # https://github.com/sqlalchemy/sqlalchemy/blob/231262d4782fc4d1a5d5fdbf985a06e04270e878/lib/sqlalchemy/sql/type_api.py#L452
```
Comment (AbdealiJK) [inline sqlalchemy_continuum/manager.py:438]: When adding a link unless you explicitly need this commit - you could just use the last stable version:
https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_40/lib/sqlalchemy/sql/type_api.py#L335
THe link becomes easier to read also.
Plus you're linking to the function there - it would be ideal to explain what you want us to read there.
For example:
```
If `result_processor` is None - no processing is required as per sqlalchemy
```

File: sqlalchemy_continuum/manager.py (hunk 2 above)
Comment (AbdealiJK) [reply sqlalchemy_continuum/manager.py:446]: In fact, now that I look at it more - isn't `parameters` a dictionary ?
Shouldn't we be doing: `parameters[column.key]` ?

File: sqlalchemy_continuum/manager.py (hunk 1 above)
Comment (indiVar0508) [reply sqlalchemy_continuum/manager.py:442]: yeah sure we can do that, kept them separate as method `get_python_processed_parameter` uses dialect to process parameters so if any dialect specific condn (process being None, not aware of any other cond's as of now) is identified the handling can be dealt within that method itself

File: sqlalchemy_continuum/manager.py (hunk 2 above)
Comment (indiVar0508) [reply sqlalchemy_continuum/manager.py:446]: yes, i think for defaults it does return some value from SQLA itself, however the issue you are pointing to i have raised same in thread for #283 for delete query perspective and i think it does exist currently in master branch also, a seperate thread is on ig for refactoring `positional_args_to_dict`

` isn't parameters a dictionary ?` - No it's a tuple `positional_args_to_dict` converts this tuple to dict

File: sqlalchemy_continuum/manager.py (hunk 3 above)
Comment (indiVar0508) [reply sqlalchemy_continuum/manager.py:438]: Sure will update this one and keep that in check for future PR(s)

**Comment by indiVar0508** [comment]
Hi @AbdealiJK @marksteward ,

Can you review changes for [address_postional_agrs issue](https://github.com/kvesteri/sqlalchemy-continuum/pull/295/commits/1ee5c41656a94d475ff9142ef15d6b9509c5d581),

i was looking into this and i think new chnages address #280 and #283 / #284, however while deleting entries i am getting #275 not sure why it seems for delete command `units_of_work` is not being populated in before_flush or before_flush is not being invoked for delete commands

pytest is working similar to latest master branch
`345 failed, 311 passed, 14 skipped, 302 warnings, 90 errors`

i was trying with below mentioned cases
```
# #280
# lotr = Book(name='Lord of the rings')
# tolkien = Author(name='JRR Tolkien', books=[lotr])
# db.add(lotr)
# db.add(tolkien)
# db.commit()
# #284/#283(PR thread)
from sqlalchemy import and_
d = book_author_table.delete(whereclause=and_(book_author_table.c.author_id==1, book_author_table.c.book_id==1))
# d = book_author_table.delete()
print(d)
engine.execute(d)
db.commit()
```
but when delete operation is triggered the statement get processed and params also seems to be correct 
`{'author_id_1': 1, 'book_id_1': 1}`

but i get a error with below mentioned traceback

```
Traceback (most recent call last):
  File "/home/indivar/github/sqlalchemy-continuum/sqlalchemy_continuum/manager.py", line 411, in append_association_operation
    uow = self.units_of_work[conn]
KeyError: <sqlalchemy.engine.base.Connection object at 0x7f5abadd7df0>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "distdemo_books.py", line 66, in <module>
    engine.execute(d)
  File "<string>", line 2, in execute
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 402, in warned
    return fn(*args, **kwargs)
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 3257, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1380, in execute
    return meth(self, multiparams, params, _EMPTY_EXECUTION_OPTS)
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/sql/elements.py", line 333, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1572, in _execute_clauseelement
    ret = self._execute_context(
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/base.py", line 1842, in _execute_context
    statement, parameters = fn(
  File "/home/indivar/github/py_envs/sqlalchemy_code_update_venv/lib/python3.8/site-packages/sqlalchemy/engine/events.py", line 128, in wrap_before_cursor_execute
    orig_fn(
  File "/home/indivar/github/sqlalchemy-continuum/sqlalchemy_continuum/manager.py", line 475, in track_association_operations
    self.append_association_operation(
  File "/home/indivar/github/sqlalchemy-continuum/sqlalchemy_continuum/manager.py", line 414, in append_association_operation
    uow = self.units_of_work[conn.engine]
KeyError: Engine(sqlite:///db.sqlite)
```

**Comment by indiVar0508** [comment]
Hi @marksteward , 

I have tidied the commits for this PR and tested the issue #282 for SQLA < 1.4, 1.3 and 1.2 it seems to works , could you review it and run travis CI Build

Pytest Run results : `346 failed, 308 passed, 14 skipped, 5398 warnings, 90 errors`

**Comment by indiVar0508** [comment]
It seems it is failing for postgres-native one, if you get a chance, could you also provide  traceback for testcase so i will try to debug the issue locally.

**Comment by indiVar0508** [comment]
Hi @marksteward ,

I have fixed the issue for last action [job run](https://github.com/kvesteri/sqlalchemy-continuum/runs/8117534892?check_suite_focus=true), the job run was failing becuase in association i missed schema name consideration for table for SQLA > 1.4, i have added that now and tested with postgres in local for failing testfile(tests/relationships/test_many_to_many_relations.py) and it seems to work now.

Could you again trigger the Travis Build?

**Comment by indiVar0508** [comment]
Hi @marksteward ,

The Travis Build seems to be success for this one, is there any other changes that are required for this PR or can accept the request and close it ?

this addresses #282 and #284

**Comment by AbdealiJK** [comment]
Hi @indiVar0508 thanks for working on this.
I was not aware of `context.compiled_parameters` - it's awesome that dwlalchemy already provides this instead of us doing all the processing we were doing earlier.

Just 2 comments:
1. Can you squash your commits as they both are solving the same thing
2. Can you add a testcase for. The scenario where an association-table has a date field ? This was we have a testcase for the reported bug.

@marksteward can this be tagged as `next-release` too ? The approach seems to be clear and this is an issue for my application

**Comment by indiVar0508** [comment]
Hi @AbdealiJK ,

Updated PR with mentioned changes

**Comment by marksteward** [comment]
Hi @indiVar0508, @AbdealiJK. Yes, I love the approach, particularly getting rid of `positional_args_to_dict`, but I want to make sure it covers any weird corner cases. Happy to add it to the next release.

**Comment by indiVar0508** [comment]
Hi @marksteward, 

sure if you want i can try to look into some corner cases and test it, can you give some inputs to point me the direction on what cases you want me to check i can try them locally
